src/order/circular.lean

```diff
@@ -0,0 +1,389 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.set.basic
+
+/-!
+# Circular order hierarchy
+
+This file defines circular preorders, circular partial orders and circular orders.
+
+## Hierarchy
+
+* A `circular_order` is a ternary betweenness relation `btw : α → α → α → Prop`, which is
+  - reflexive: `btw a a a`
+  - cyclic: `btw a b c → btw b c a`
+  - antisymmetric: `btw a b c → btw c b a → a = b ∨ b = c ∨ c = a`
+  - total: `btw a b c ∨ btw c b a`
+  along with a strict betweenness relation `sbtw : α → α → α → Prop` which respects
+  `sbtw a b c ↔ btw a b c ∧ ¬ btw c b a`, analogously to how `<` and `≤` are related, and is
+  - transitive: `sbtw a b c → sbtw b d c → sbtw a d c`.
+* A `circular_partial_order` drops totality.
+* A `circular_preorder` further drops antisymmetry.
+
+The mind picture is that a circular order is a circle and `btw a b c` means that going around
+clockwise from `a` you reach `b` before `c` (`b` is between `a` and `c` is meaningless on an
+unoriented circle). A circular partial order is several, potentially intersecting, circles. A
+circular preorder is more of a mess.
+
+One can translate from usual orders to circular ones by "closing the necklace at infinity". See
+`has_le.to_has_btw` and `has_lt.to_has_sbtw`. Going the other way involves "cutting the necklace" or
+"rolling the necklace open".
```

Comment:
Do you already have machinery for this direction?

Reply:
Cutting the necklace is trivial. Rolling it open (which means putting an order on `α × ℤ`) isn't hard either but requires some thought. Eg, what's the correct generality? What about `ℤ × α`?

Reply:
Sure, so maybe briefly mention that this is TODO